cuda/solver/upper_trs_kernels.cu

```diff
@@ -63,19 +64,99 @@ namespace upper_trs {
 
 
 void should_perform_transpose(std::shared_ptr<const CudaExecutor> exec,
-                              bool &do_transpose) GKO_NOT_IMPLEMENTED;
+                              bool &do_transpose)
+{
+#if (defined(CUDA_VERSION) && (CUDA_VERSION >= 9020))
+
+
+    do_transpose = false;
+
+
+#elif (defined(CUDA_VERSION) && (CUDA_VERSION < 9020))
+
+
+    do_transpose = true;
+
+
+#endif
+}
 
 
 void init_struct(std::shared_ptr<const CudaExecutor> exec,
                  std::shared_ptr<solver::SolveStruct> &solve_struct)
-    GKO_NOT_IMPLEMENTED;
+{
+    solve_struct =
+        std::shared_ptr<solver::SolveStruct>(new solver::SolveStruct());
+}
 
 
 template <typename ValueType, typename IndexType>
 void generate(std::shared_ptr<const CudaExecutor> exec,
               const matrix::Csr<ValueType, IndexType> *matrix,
-              solver::SolveStruct *solve_struct,
-              const gko::size_type num_rhs) GKO_NOT_IMPLEMENTED;
+              solver::SolveStruct *solve_struct, const gko::size_type num_rhs)
+{
+    if (cusparse::is_supported<ValueType, IndexType>::value) {
```

Comment:
You probably are missing `device_guard` calls pretty much everywhere in your code. Including in the previous things you committed. I wasn't careful about that.